discord_bot/discord_bot.py

```diff
@@ -34,53 +36,94 @@ def load_config():
         with open(CONFIG_FILE_PATH) as f:
             return json.load(f)
 
-    def get_gateway(self):
+    def get_gateway(self, **kwargs):
         """
-        Caches a gateway value and retrieves a new URL
+        Caches a gateway value, authenticates, and retrieves a new URL
         :return: gateway URL
         """
-        r = requests.get(self.config["discord_gateway_endpoint"])
+        header = {
+            "headers": {
+                "Authorization": "Bot {}".format(self.config["handshake_identity"]["token"]),
+                "User-Agent": "DiscordBot (https://github.com/ShifuYee/Discord-Bot, 0.0.1)"
+            }
+        }
+        kwargs = dict(header, **kwargs)
+        r = requests.get(self.config["discord_gateway_endpoint"], **kwargs)
+        assert 200 == r.status_code, r.reason
         return r.json()['url']
 
     async def send_json(self, payload):
         asyncio.ensure_future(self.websocket.send(json.dumps(payload)))
 
     async def handshake(self, message):
         """
-        When a websocket connection is opened, Hello payload is received. Then, an Identify payload is sent as
+        When a websocket connection is opened, Hello payload is received. Then, an Identify/Resume payload is sent as
         part of the handshake to authorize this client.
 
         :param message: dict - hello payload
         """
-        self.heartbeat_interval_ms = message["d"]["heartbeat_interval"]
+        self.heartbeat_interval_ms = message["d"]["heartbeat_interval"]  # seconds
         asyncio.ensure_future(self.heartbeat())
-
         handshake_identity = self.config["handshake_identity"]
-        asyncio.ensure_future(self.send_json({"op": Opcodes.IDENTIFY, "d": handshake_identity}))
+        if not self.session_id:
+            asyncio.ensure_future(self.send_json({"op": Opcodes.IDENTIFY, "d": handshake_identity}))
+        else:
+            asyncio.ensure_future(self.send_json({"op": Opcodes.RESUME,
+                                                  "d": {"token": self.config["handshake_identity"]["token"],
+                                                        "session_id": self.session_id,
+                                                        "seq": self.last_seq}}))
 
     async def heartbeat(self):
         """
-        Sends a heartbeat payload every heartbeat interval.
+        Sends a heartbeat payload every heartbeat interval and allows messages to be sent.
         """
         await asyncio.sleep(self.heartbeat_interval_ms / 1000.0)
+        print("Last Sequence: {}".format(self.last_seq))
         asyncio.ensure_future(self.send_json({"op": Opcodes.HEARTBEAT, "d": self.last_seq}))
         asyncio.ensure_future(self.heartbeat())
+        # asyncio.ensure_future(self.send_message("embed": {
```

Comment:
This should probably go in a separate method, even for testing purposes. Also, all temporary code should at least have a `# TODO: Remove this test code` above it. Things marked `# TODO` get special highlighting.